### PR TITLE
nm.checkpoint: drop dependency on python-dbus

### PR DESCRIPTION
**Pull request by ffmancera**
Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

### CODE REVIEW

**Comment by ffmancera** [comment]
This PR is based on #236 from @cathay4t. I couldn't rebase or cherry-pick it because there were too many conflicts. If he agrees, I will add his "Signed-off-by". Thanks!

**Comment by cathay4t** [comment]
And the CI is failling due to some checkpoint issue. Please investigate.

**Comment by ffmancera** [comment]
> And the CI is failling due to some checkpoint issue. Please investigate.

Yes I found out the issue. It is failing with the new MainContext, fixing it. Thanks!

**Comment by ffmancera** [comment]
I found an issue in ActiveConnection. Fixing it. It is required to continue with this patch.

File: libnmstate/nm/checkpoint.py [hunk 1]
```diff
@@ -112,26 +70,32 @@ def create(self):
...
+        self._ctx.register_async("Create checkpoint")
         try:
-            dbuspath = self._manager.interface.CheckpointCreate(
-                devs, timeout, cp_flags
+            self._ctx.client.checkpoint_create(
+                devs,
+                timeout,
+                cp_flags,
+                self._ctx.cancellable,
+                CheckPoint._checkpoint_create_callback,
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:85]: Accessing private member is not suggested. Use `self` here.

File: libnmstate/nm/checkpoint.py [hunk 2]
```diff
@@ -149,31 +113,126 @@ def _refresh_checkpoint_timeout(self, _user_data):
...
+            checkpoint._ctx.fail(
+                NmstateLibnmError(f"Checkpoint rollback failed: error={e}")
+            )
+
+    @staticmethod
+    def _check_rollback_result(client, result, dbus_path):
+        ret = client.checkpoint_rollback_finish(result)
+        logging.debug(f"Checkpoint {dbus_path} rollback executed")
+        for path in ret:
+            d = client.get_device_by_path(path)
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:215]: Please use nm_dev instead.

**Comment by ffmancera** [comment]
@cathay4t I've found a way to do this using the signal from nm_dev. I am testing it and will be ready by tomorrow morning. Thanks!

File: libnmstate/nm/checkpoint.py (hunk 2 above)
Comment (ffmancera) [reply libnmstate/nm/checkpoint.py]: Done, thanks!

File: libnmstate/nm/checkpoint.py (hunk 1 above)
Comment (ffmancera) [reply libnmstate/nm/checkpoint.py]: Done!

**Comment by ffmancera** [comment]
@cathay4t Packit build is failing due to a major issue in Copr. It is affecting everyone.

File: libnmstate/nm/checkpoint.py [hunk 3]
```diff
@@ -149,31 +116,145 @@ def _refresh_checkpoint_timeout(self, _user_data):
...
+    def _check_rollback_result(self, client, result, dbus_path):
+        ret = client.checkpoint_rollback_finish(result)
+        logging.debug(f"Checkpoint {dbus_path} rollback executed")
+        for path in ret:
+            nm_dev = client.get_device_by_path(path)
+            iface = path if nm_dev is None else nm_dev.get_iface()
+            if (
+                nm_dev.get_state() == common.NM.DeviceState.DEACTIVATING
+                and nm_dev.get_state_reason()
+                == common.NM.DeviceStateReason.NEW_ACTIVATION
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:217]: Device might in `NM.DeviceState.IP_CONFIG` which also need wait.

File: libnmstate/nm/checkpoint.py [hunk 4]
```diff
@@ -149,31 +116,145 @@ def _refresh_checkpoint_timeout(self, _user_data):
...
+            ):
+                action = f"Waiting for rolling back {iface}"
+                self._ctx.register_async(action)
+                handler_id = nm_dev.connect(
+                    NM_DEV_STATE_CHANGED_SIGNAL,
+                    self._wait_nm_dev_ready_callback,
+                    action,
+                )
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:225]: Please find a way to unify this action with code in `nm/active_connection.py` and `nm/connection.py`.

File: libnmstate/nm/checkpoint.py [hunk 5]
```diff
@@ -149,31 +116,145 @@ def _refresh_checkpoint_timeout(self, _user_data):
...
+            else:
+                logging.debug(f"Interface {iface} rollback succeeded")
+
+    def _wait_nm_dev_ready_callback(
+        self, nm_dev, state, last_state, reason, action
+    ):
+        if self._ctx.is_cancelled():
+            print("CANCELED!!")
+            return
+        if state == common.NM.DeviceState.ACTIVATED:
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:238]: This is incorrect.
We should follow the same rule on activation: don't wait on master interface, don't wait DHCP, etc.

Please find a way to unify this action with code in nm/active_connection.py and nm/connection.py.

File: libnmstate/nm/checkpoint.py [hunk 6]
```diff
@@ -149,31 +116,145 @@ def _refresh_checkpoint_timeout(self, _user_data):
...
+            print("ACTIVATED")
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:239]: This is leftover.

File: libnmstate/nm/checkpoint.py (hunk 3 above)
Comment (ffmancera) [reply libnmstate/nm/checkpoint.py:218]: Done, thanks!

File: libnmstate/nm/checkpoint.py (hunk 4 above)
Comment (ffmancera) [reply libnmstate/nm/checkpoint.py]: I unified this with `nm/connection.py` using `wait_dev_activation()`.

File: libnmstate/nm/checkpoint.py (hunk 5 above)
Comment (ffmancera) [reply libnmstate/nm/checkpoint.py]: I agree. Look at the new approach. Thanks!

File: libnmstate/nm/checkpoint.py (hunk 6 above)
Comment (ffmancera) [reply libnmstate/nm/checkpoint.py]: Ooops, leftovers removed. Thanks!

File: libnmstate/nm/checkpoint.py [hunk 7]
```diff
@@ -149,31 +117,127 @@ def _refresh_checkpoint_timeout(self, _user_data):
...
+                )
+                self._ctx.finish_async("Create checkpoint")
+            else:
+                error_msg = (
+                    f"dbuspath={self._dbuspath} "
+                    f"timeout={self._timeout} "
+                    f"callback result={cp}"
+                )
+                self._ctx.fail(
+                    NmstateLibnmError(f"Checkpoint create failed: {error_msg}")
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:191]: Should handle two cases:
 * NmstateConflictError for two+ checkpoint.
 * NmstatePermissionError for non-root.

Comment (ffmancera) [reply libnmstate/nm/checkpoint.py:193]: Done, this is now handing the permission and the conflict error on the try/except.

File: libnmstate/nm/checkpoint.py [hunk 8]
```diff
@@ -19,18 +19,14 @@
...
 
-
-DBUS_STD_PROPERTIES_IFNAME = "org.freedesktop.DBus.Properties"
-
-CHECKPOINT_CREATE_FLAG_DESTROY_ALL = 0x01
-CHECKPOINT_CREATE_FLAG_DELETE_NEW_CONNECTIONS = 0x02
-CHECKPOINT_CREATE_FLAG_DISCONNECT_NEW_DEVICES = 0x04
-
-NM_PERMISSION_DENIED = "org.freedesktop.NetworkManager.PermissionDenied"
+NM_DEV_STATE_CHANGED_SIGNAL = "state-changed"
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:29]: This is left over.

File: libnmstate/nm/checkpoint.py [hunk 9]
```diff
@@ -86,11 +51,11 @@ def __init__(
         self, nm_context, autodestroy=True, timeout=60, dbuspath=None
     ):
         self._ctx = nm_context
-        self._manager = nmdbus_manager()
         self._timeout = timeout
         self._dbuspath = dbuspath
         self._timeout_source = None
         self._autodestroy = autodestroy
+        self._handlers = {}
```
Comment (cathay4t) [inline libnmstate/nm/checkpoint.py:58]: This is leftover.

**Comment by cathay4t** [comment]
The CI still fails on `test_disable_static_ipv6_and_rollback`

**Pull request merged**